includes/sanitizers/class-amp-core-theme-sanitizer.php

```diff
@@ -1515,4 +1545,293 @@ public function add_twentyfourteen_search() {
+
+			// If we got back an existing ID, then the toggle target was already targeted by a previous toggle,
+			//so we don't need to adapt the target, only the toggle (multiple toggles can target the same target node).
+			if ( ! array_key_exists( $id, $state ) ) {
+				$state[ $id ] = [
+					'on'           => $on_class,
+					'off'          => $off_class,
+					'css_selector' => $toggle_target,
+				];
+
+				// We can't use the regular "[class]" notation because of the PHP dom extension.
+				$target_node->setAttribute( 'data-amp-bind-class', "{$id} ? {$id}_classes['on'] : {$id}_classes['off']" );
+			}
+
+			$this->add_amp_action( $toggle, 'tap', "AMP.setState({{$id}: !{$id}})" );
+		}
+
+		// Add <amp-state> snippets to the document that contain the classes to use.
+		foreach ( $state as $id => $state_data ) {
+			$amp_state    = $this->dom->createElement( 'amp-state' );
+			$amp_state->setAttribute( 'id', "{$id}_classes" );
+			unset( $state_data['css_selector'] );
+			$script = $this->dom->createElement( 'script' );
+			$script->setAttribute( 'type', 'application/json' );
+			$script->appendChild( $this->dom->createTextNode( wp_json_encode( $state_data ) ) );
+			$amp_state->appendChild( $script );
+			$this->body->appendChild( $amp_state );
+		}
+
+	}
+
+	/**
+	 * Get the next toggle ID, or return a previous one for a same selector.
+	 *
+	 * @param string $css_selector Selector to get the toggle ID for.
+	 * @param array  $state        Associative array of existing state to check.
+	 * @return string Toggle ID to use.
+	 */
+	protected function get_toggle_id( $css_selector, $state ) {
+		static $index = 1;
+
+		if ( 'next' !== $css_selector ) {
+			foreach ( $state as $id => $state_data ) {
+				if ( $state_data['css_selector'] === $css_selector ) {
+					return $id;
+				}
+			}
+		}
+
+		$id = "_amp_toggle_{$index}";
+		$index ++;
+
+		return $id;
+	}
+
+	/**
+	 * Provides a "best guess" as to what XPath would mirror a given CSS
+	 * selector.
+	 *
+	 * This is a very simplistic conversion and will only work for very basic
+	 * CSS selectors. Therefore, it provides a filter for themes & plugins to
+	 * hook into to provide custom replacements.
+	 *
+	 * @param string $css_selector CSS selector to convert.
+	 * @return string|null XPath that closely mirrors the provided CSS
+	 *                     selector, or null if an error occurred.
+	 */
+	protected function xpath_from_css_selector( $css_selector ) {
+		// Start with basic clean-up.
+		$css_selector = trim( $css_selector );
+		$css_selector = preg_replace( '/\s+/', ' ', $css_selector );
+
+		/**
+		 * Provide a manual conversion from CSS selector to XPath query.
+		 *
+		 * @param string CSS selector that needs to be converted to XPath.
+		 * @return string XPath to use. Return the unchanged CSS selector if no conversion provided.
+		 */
+		$xpath = apply_filters( 'amp_xpath_from_css_selector', $css_selector );
```

Comment:
Using a filter here will be unreliable because it will make the post-processed output no longer deterministic based on the args being passed to the sanitizers.

Do we need to use XPath? See also #2102.

Reply:
Ah, yes, of course. I'll remove the filter again.

Also, great find with that Symfony package, that would come in handy. Would you consider it an option to use a quick 5.6 fork until we can use the real thing? Getting from 7.1 to 5.6 shouldn't be much more than stripping the return types for most of the code. Maybe it can even be done automatically with https://github.com/rectorphp/rector

If we have that fork available, we could also start looking into improving the tree shaker.